src/snovault/elasticsearch/searches/configs.py

```diff
@@ -46,3 +55,188 @@ def __init__(self, allowed_kwargs=[], **kwargs):
             allowed_kwargs=allowed_kwargs or DEFAULT_EXISTS_AGGREGATION_KWARGS,
             **kwargs
         )
+
+
+class SortedTupleMap:
+
+    def __init__(self):
+        self._map = {}
+
+    @staticmethod
+    def _convert_key_to_sorted_tuple(key):
+        if isinstance(key, str):
+            key = [key]
+        return tuple(sorted(key))
+
+    def __setitem__(self, key, value):
+        key = self._convert_key_to_sorted_tuple(key)
+        self._map[key] = value
+
+    def __getitem__(self, key):
+        key = self._convert_key_to_sorted_tuple(key)
+        return self._map[key]
+
+    def __contains__(self, key):
+        key = self._convert_key_to_sorted_tuple(key)
+        return key in self._map
+
+    def get(self, key, default=None):
+        return self._map.get(
+            self._convert_key_to_sorted_tuple(key),
+            default
+        )
+
+    def drop(self, key):
+        key = self._convert_key_to_sorted_tuple(key)
+        if key in self._map:
+            del self._map[key]
+
+    def as_dict(self):
+        return dict(self._map)
+
+
+def get_search_config():
+    return SearchConfig
+
+
+class SearchConfigRegistry:
+
+    def __init__(self):
+        self._initialize_storage()
+
+    def _initialize_storage(self):
+        self.registry = SortedTupleMap()
+        self.aliases = SortedTupleMap()
+        self.defaults = SortedTupleMap()
+
+    def add(self, config):
+        self.registry[config.name] = config
+
+    def add_aliases(self, aliases):
+        for k, v in aliases.items():
+            self.aliases[k] = v
+
+    def add_defaults(self, defaults):
+        for k, v in defaults.items():
+            self.defaults[k] = v
+
+    def update(self, config):
+        if config.name in self.registry:
+            self.get(config.name).update(**config)
+        else:
+            self.add(config)
+
+    def register_from_func(self, name, func):
+        config = get_search_config()(name, func())
+        self.update(config)
+
+    def register_from_item(self, item):
+        config = get_search_config().from_item(item)
+        self.update(config)
+
+    def clear(self):
+        self._initialize_storage()
+
+    def get(self, name, default=None):
+        return self.registry.get(name, default)
+
+    def _resolve_config_name(self, name, use_defaults=True):
+        if name in self.aliases:
+            yield from self._resolve_config_names(
+                self.aliases[name],
+                use_defaults=use_defaults
+            )
+        elif use_defaults and name in self.defaults:
+            yield from self._resolve_config_names(
```

Comment:
if name == self.defaults[name] or self.aliases[name] we will get an infinite loop. It's something we can control in the definitions of the config dict, but should we add something preventing that to be set? this is not a blocker for now...